stl/inc/xutility

```diff
@@ -4198,7 +4198,368 @@ _NODISCARD _CONSTEXPR17 move_iterator<_Iter> make_move_iterator(_Iter _It) { //
+        if (_Left._Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_iter) {
+            return _Left._Myiter == _Right._Myiter;
+        } else if (_Left._Mystate == _State::_Holds_sentinel && _Right._Mystate == _State::_Holds_sentinel) {
+            return true;
+        } else if (_Left._Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_sentinel) {
+            return _Left._Myiter == _Right._Mysentinel;
+        } else {
+            return _Left._Mysentinel == _Right._Myiter;
+        }
+    }
+
+    // clang-format off
+    template <sized_sentinel_for<_Iter> _Iter2, sized_sentinel_for<_Iter> _Se2>
+        requires sized_sentinel_for<_Se, _Iter2>
+    _NODISCARD friend iter_difference_t<_Iter2> operator-(
+        const common_iterator& _Left, const common_iterator<_Iter2, _Se2>& _Right) {
+        // clang-format on
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Left._Mystate != _State::_Nothing && _Right._Mystate != _State::_Nothing,
+            "Cannot take difference of valueless common_iterators");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if (_Left._Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_iter) {
+            return _Left._Myiter - _Right._Myiter;
+        } else if (_Left._Mystate == _State::_Holds_sentinel && _Right._Mystate == _State::_Holds_sentinel) {
+            return 0;
+        } else if (_Left._Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_sentinel) {
+            return _Left._Myiter - _Right._Mysentinel;
+        } else {
+            return _Left._Mysentinel - _Right._Myiter;
+        }
+    }
+
+    // clang-format off
+    _NODISCARD friend iter_rvalue_reference_t<_Iter> iter_move(const common_iterator& _Right) noexcept(
+        noexcept(_RANGES iter_move(_STD declval<const _Iter&>())))
+        requires input_iterator<_Iter>
+    {
+        // clang-format on
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(
+            _Right._Mystate == _State::_Holds_iter, "Can only iter_move from common_iterator if it holds an iterator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return _RANGES iter_move(_Right._Myiter);
+    }
+
+    template <indirectly_swappable<_Iter> _Iter2, class _Se2>
+    friend void iter_swap(const common_iterator& _Left, const common_iterator<_Iter2, _Se2>& _Right) noexcept(
+        noexcept(_RANGES iter_swap(_STD declval<const _Iter&>(), _STD declval<const _Iter2&>()))) {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Left._Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_iter,
+            "Can only iter_swap common_iterators if both hold an iterator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return _RANGES iter_swap(_Left._Myiter, _Right._Myiter);
+    }
+
+private:
+    enum class _State : unsigned char { _Nothing, _Holds_iter, _Holds_sentinel };
+
+    union {
+        _Iter _Myiter;
+        _Se _Mysentinel;
+        _Nontrivial_dummy_type _Mydummy;

```

Comment:
IIUC, `std::optional` needs `_Nontrivial_dummy_type ` because we don't want `std::optional<int> o{};` to zero out the contained value, but here we do want the default constructor to zero out `_Myiter`, so I think this member is not needed.

Reply:
This is not for the default constructor where we indeed create a default initialized `_Myiter` but for the copy constructors.

There we do not know before hand whether `_Right` will hold an iterator or an sentinel. So in that case we initialize to `_Nontrivial_dummy_type` which should be an noop.

Does that make sense?

Reply:
I think you can just omit the member initializer in that case.

Reply:
I am unsure about that. At the end of the initializer list the object must be fully initialized. I do not know what the compiler does with a union without a default member. As I said `union` scares me